### PR TITLE
Go Driver feedback

### PR DESCRIPTION
**Pull request by justing-bq**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: qldbdriver/qldbdriver.go [hunk 1]
```diff
@@ -69,24 +69,23 @@ func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*D
 	semaphore := sync2.NewSemaphore(int(options.MaxConcurrentTransactions), 0)
 	sessionPool := make(chan *session, options.MaxConcurrentTransactions)
 	isClosed := false
+	retryPolicy := RetryPolicy{MaxRetryLimit: 4, Backoff: ExponentialBackoffStrategy{SleepBase: 10000000, SleepCap: 5000000000}}
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:72]: You can't just give the default.

This needs to be set as the default in the driver options, and then grabbed from the options struct if a user wants to modify it. Also, are we not using time durations for sleep base and sleep cap?

File: qldbdriver/qldbdriver.go [hunk 2]
```diff
@@ -69,24 +69,23 @@ func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*D
...
+// SetRetryPolicy sets the driver's retry policy for Execute
+func (driver *QLDBDriver) SetRetryPolicy(rp RetryPolicy) {
+	driver.retryPolicy = rp
 }
 
-// Execute a provided function within the context of a new QLDB transaction with a custom RetryPolicy.
+// Execute a provided function within the context of a new QLDB transaction.
 //
 // The provided function might be executed more than once.
+// This function is not expected to ever run concurrently.
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:86]: I would say that the provided function is not expected to run concurrently. 

Your current wording implies that Execute isn't supporting concurrency which is wrong.

File: qldbdriver/qldbdriver.go [hunk 3]
```diff
@@ -122,25 +121,25 @@ func (driver *QLDBDriver) ExecuteWithRetryPolicy(ctx context.Context, fn func(tx
 			}
 			// Retry
 			retryAttempt++
-			driver.logger.log(fmt.Sprintf("A recoverable error has occurred. Attempting retry #%d.", retryAttempt), LogInfo)
-			driver.logger.log(fmt.Sprintf("Errored Transaction ID: %s. Error cause: %v", txnErr.transactionID, txnErr), LogDebug)
+			driver.logger.logf(LogInfo, "A recoverable error has occurred. Attempting retry #%d.", retryAttempt)
+			driver.logger.logf(LogDebug, "Errored Transaction ID: %s. Error cause: '%v'", txnErr.transactionID, txnErr)
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:125]: why '%v' here, but not '%s'? What are the single quotes for?

File: qldbdriver/qldbdriver.go [hunk 4]
```diff
@@ -184,33 +183,33 @@ func (driver *QLDBDriver) GetTableNames(ctx context.Context) ([]string, error) {
...
+				driver.logger.logf(LogDebug, "Encountered error trying to end session: '%v'", err.Error())
 			}
 		}
 		close(driver.sessionPool)
 	}
 }
 
 func (driver *QLDBDriver) getSession(ctx context.Context) (*session, error) {
-	driver.logger.log(fmt.Sprint("Getting session. Existing sessions available:", len(driver.sessionPool)), LogDebug)
+	driver.logger.logf(LogDebug, "Getting session. Existing sessions available: %v", len(driver.sessionPool))
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:200]: This seems inconsistent again. Can you double check the single quoting of logging in this pr

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: qldbdriver/qldbdriver.go (hunk 3 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go:125]: %v can be used for any type and would behave the same as %s for strings.
The single quotes are just to enclose the error message.

File: qldbdriver/qldbdriver.go (hunk 4 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go:200]: No single quotes here because it's just a number.

File: qldbdriver/qldbdriver.go (hunk 1 above)
Comment (justing-bq) [reply qldbdriver/qldbdriver.go:72]: SleepBase and SleepCap are now time.Duration and when assigned to a number type, the number is interpreted as nano seconds hence why they are set to 10000000 and 5000000000 respectively.

File: qldbdriver/qldbdriver.go [hunk 5]
```diff
@@ -70,23 +71,20 @@ func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*D
...
+		semaphore, sessionPool, retryPolicy}
 }
 
-// Execute a provided function within the context of a new QLDB transaction.
-//
-// The provided function might be executed more than once.
-// It is recommended for it to be idempotent, so that it doesn't have unintended side effects in the case of retries.
-func (driver *QLDBDriver) Execute(ctx context.Context, fn func(txn Transaction) (interface{}, error)) (interface{}, error) {
-	return driver.ExecuteWithRetryPolicy(ctx, fn, RetryPolicy{MaxRetryLimit: 4, Backoff: ExponentialBackoffStrategy{SleepBaseInMillis: 10, SleepCapInMillis: 5000}})
+// SetRetryPolicy sets the driver's retry policy for Execute
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:78]: Nit: missing period

File: qldbdriver/qldbdriver.go [hunk 6]
```diff
@@ -27,10 +26,10 @@ import (
 
 // DriverOptions can be used to configure the driver during construction.
 type DriverOptions struct {
-	// The amount of times a transaction will automatically retry upon a recoverable error. Default: 4.
-	RetryLimit uint8
+	// The policy guiding retry attempts upon a recoverable error. Default: MaxRetryLimit: 4, SleepBase: 10ms, SleepCap: 5000 ms.
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:29]: Nit: SleepCap has an extra space after 5000.

Also, are we not mentioning the default is exponential?

File: qldbdriver/qldbdriver.go [hunk 7]
```diff
@@ -70,23 +71,20 @@ func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*D
 	sessionPool := make(chan *session, options.MaxConcurrentTransactions)
 	isClosed := false
 
-	return &QLDBDriver{ledgerName, qldbSession, options.RetryLimit, options.MaxConcurrentTransactions, logger, isClosed,
-		semaphore, sessionPool}
+	return &QLDBDriver{ledgerName, qldbSession, options.RetryPolicy.MaxRetryLimit, options.MaxConcurrentTransactions, logger, isClosed,
+		semaphore, sessionPool, retryPolicy}
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:75]: This needs to be options.retryPolicy if the user enters their own

File: qldbdriver/qldbdriver.go [hunk 8]
```diff
@@ -41,20 +40,22 @@ type DriverOptions struct {
...
+	retryPolicy               RetryPolicy
 }
 
 // New creates a QLBDDriver using the parameters and options, and verifies the configuration.
 //
 // Note that qldbSession.Client.Config.MaxRetries will be set to 0. This property should not be modified.
 // DriverOptions.RetryLimit is unrelated to this property, but should be used if it is desired to modify the amount of retires for statement executions.
 func New(ledgerName string, qldbSession *qldbsession.QLDBSession, fns ...func(*DriverOptions)) *QLDBDriver {
-	options := &DriverOptions{4, 50, defaultLogger{}, LogInfo}
+	retryPolicy := RetryPolicy{MaxRetryLimit: 4, Backoff: ExponentialBackoffStrategy{SleepBase: 10000000, SleepCap: 5000000000}}
```
Comment (simonz-bq) [inline qldbdriver/qldbdriver.go:57]: To clarify what I meant earlier, you need to say SleepBase: 10000 * Time.Duration.Milliseconds in order to specify what's happening here.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

**Review by simonz-bq — APPROVED** [review]

**Review by yoroto — APPROVED** [review]

**Pull request merged**